modules/maps.js

```diff
@@ -15,207 +15,770 @@ MODULES.maps.SkipNumUnboughtPrestiges = 2;
+        name: profile.name,
+        minLvl: profile.minLevel,
+        baseLvl: profile.baseLvl,
+        optimalLvl: profile.optimalLevel,
+        biome: profile.preferredBiome,
+        mods: `[${profile.mods}]`,
+        required: `[${profile.required}]`,
+        optional: `[${profile.optional}]`,
+    }, '=', true);
+
+    // reset all map params to the lowest (cheapest) setting
+    resetAdvMaps(true);
+    setMapLevel(profile.minLevel);
+
+    // first set all required properties of the map
+    for (const req of profile.required) {
+        if (mapSliders.includes(req)) {
+            setMapSlider(req, 9);
+        } else if (mapLevels.includes(req)) {
+            setMapLevel(profile.getTargetLevel(req));
+        } else if (req === MappingProfile.priorities.biome) {
+            setMapBiome(profile.preferredBiome);
+        } else if (req === MappingProfile.priorities.mod && profile.mods.length) {
+            setMapMod(profile.mods[0]);
+        } else {
+            devDebug(ctx, 'Unknown map requirement', {req: req}, '=', true);
+        }
+    }
+    if (!canAffordSelectedMap()) {
+        fragmentsNeeded = updateMapCost(true);
+        if (currentMap) {
+            // continue running an existing acceptable map until we can afford an upgrade
+            return false;
+        } else {
+            // we need to craft any affordable map to run
+            setAffordableMapLevel(profile.minLevel, profile.optimalLevel);
+            setAffordableMapMod(profile.mods);
+            return canAffordSelectedMap() && shouldBuyNewMap(highestMap, profile.optimalLevel);
+        }
+    }
+
+    // now we can try to improve optional parameters
+    for (const opt of profile.optional) {
+        if (mapSliders.includes(opt)) {
+            let value = 9;
+            setMapSlider(opt, value);
+            if (canAffordSelectedMap()) {
+                // improve the next option
+                continue;
+            }
+            fragmentsNeeded = Math.max(fragmentsNeeded, updateMapCost(true));
+            // gradually decrement slider until we can afford it
+            while (!canAffordSelectedMap() && value > 0) {
+                value -= 1;
+                setMapSlider(opt, value);
+            }
+            break; // can't afford this option, no point checking further
+        } else if (mapLevels.includes(opt)) {
+            let level = profile.getTargetLevel(opt);
+            setMapLevel(level);
+            if (canAffordSelectedMap()) {
+                // improve the next option
+                continue;
+            }
+            fragmentsNeeded = Math.max(fragmentsNeeded, updateMapCost(true));
+            // gradually decrement slider until we can afford it
+            while (!canAffordSelectedMap() && level > profile.minLevel) {
+                level -= 1;
+                setMapLevel(level);
+            }
+            break; // can't afford this option, no point checking further
```

Comment:
This is where the bug with neededFragments is coming from.

First. it tries to max the map level up to baseMapLvl or optimalMapLevel.
Then, it decreases that level until AT can afford it.
However, it keeps the value found at the first (and highest) map level it tried thanks to the Math.max() function.